### PR TITLE
Allow updating additional_capabilities

### LINKED ISSUE #7556: Changing ultra_ssd_enabled should not force recreation
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

* `Terraform v0.12.24`
* `terraform-provider-azurerm_v2.16.0_x5`

### Affected Resource(s)

* `azurerm_linux_virtual_machine`
* `azurerm_windows_virtual_machine`

### Terraform Configuration Files

```hcl
resource "azurerm_linux_virtual_machine" "instance" {

  [...]

  additional_capabilities {
    ultra_ssd_enabled = false
  }
}

```

### Expected Behavior

Changing `ultra_ssd_enabled` should be possible without recreation if the VM is stopped/deallocated.

### Actual Behavior

Changing `ultra_ssd_enabled` requires recreation of the VM resource also if the VM is stopped/deallocated. This possible using Azure CLI as well.

### Steps to Reproduce

1. Create configuration for a virtual machine as described above
2. `terraform apply`
3. Change `ultra_ssd_enabled` to `true`
4. `terraform plan/apply`

### PR DESCRIPTION
**Pull request by NilsBusche**
Add update for additional_capabilities.

Allow updating ultra_ssd_enabled without recreation as mentioned in https://github.com/terraform-providers/terraform-provider-azurerm/issues/7556

Fixes #7556

### CODE REVIEW

File: azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go [hunk 1]
```diff
@@ -542,6 +542,82 @@ func TestAccWindowsVirtualMachine_otherTimeZone(t *testing.T) {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { acceptance.PreCheck(t) },
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: checkWindowsVirtualMachineIsDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: testWindowsVirtualMachine_otherUltraSsdDefault(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkWindowsVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.ultra_ssd_enabled", "false"),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go:557]:
```suggestion
					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.0.ultra_ssd_enabled", "false"),
```

File: azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go [hunk 2]
```diff
@@ -542,6 +542,82 @@ func TestAccWindowsVirtualMachine_otherTimeZone(t *testing.T) {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { acceptance.PreCheck(t) },
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: checkWindowsVirtualMachineIsDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: testWindowsVirtualMachine_otherUltraSsdEnabled(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkWindowsVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.ultra_ssd_enabled", "true"),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go:579]: this is a List so needs to be:

```suggestion
					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.0.ultra_ssd_enabled", "true"),
```

File: azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go [hunk 3]
```diff
@@ -354,6 +354,74 @@ func TestAccLinuxVirtualMachine_otherTags(t *testing.T) {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { acceptance.PreCheck(t) },
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: checkLinuxVirtualMachineIsDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: testLinuxVirtualMachine_otherUltraSsdDefault(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkLinuxVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.ultra_ssd_enabled", "false"),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:369]:
```suggestion
					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.0.ultra_ssd_enabled", "false"),
```

File: azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go [hunk 4]
```diff
@@ -354,6 +354,74 @@ func TestAccLinuxVirtualMachine_otherTags(t *testing.T) {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { acceptance.PreCheck(t) },
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: checkLinuxVirtualMachineIsDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: testLinuxVirtualMachine_otherUltraSsdEnabled(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkLinuxVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.ultra_ssd_enabled", "true"),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:389]:
```suggestion
					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.0.ultra_ssd_enabled", "true"),
```

File: azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go [hunk 5]
```diff
@@ -354,6 +354,74 @@ func TestAccLinuxVirtualMachine_otherTags(t *testing.T) {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { acceptance.PreCheck(t) },
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: checkLinuxVirtualMachineIsDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: testLinuxVirtualMachine_otherUltraSsdDefault(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkLinuxVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.ultra_ssd_enabled", "false"),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:409]:
```suggestion
					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.0.ultra_ssd_enabled", "false"),
```

File: azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go [hunk 6]
```diff
@@ -354,6 +354,74 @@ func TestAccLinuxVirtualMachine_otherTags(t *testing.T) {
...
+				),
+			},
+			data.ImportStep(),
+			{
+				Config: testLinuxVirtualMachine_otherUltraSsdEnabled(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkLinuxVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.ultra_ssd_enabled", "true"),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go:417]:
```suggestion
					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.0.ultra_ssd_enabled", "true"),
```

File: azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go [hunk 7]
```diff
@@ -542,6 +542,82 @@ func TestAccWindowsVirtualMachine_otherTimeZone(t *testing.T) {
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:     func() { acceptance.PreCheck(t) },
+		Providers:    acceptance.SupportedProviders,
+		CheckDestroy: checkWindowsVirtualMachineIsDestroyed,
+		Steps: []resource.TestStep{
+			{
+				Config: testWindowsVirtualMachine_otherUltraSsdDefault(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkWindowsVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.ultra_ssd_enabled", "false"),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go:601]:
```suggestion
					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.0.ultra_ssd_enabled", "false"),
```

File: azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go [hunk 8]
```diff
@@ -542,6 +542,82 @@ func TestAccWindowsVirtualMachine_otherTimeZone(t *testing.T) {
...
+				),
+			},
+			data.ImportStep(
+				"admin_password",
+			),
+			{
+				Config: testWindowsVirtualMachine_otherUltraSsdEnabled(data),
+				Check: resource.ComposeTestCheckFunc(
+					checkWindowsVirtualMachineExists(data.ResourceName),
+					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.ultra_ssd_enabled", "true"),
```
Comment (tombuildsstuff) [inline azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go:611]:
```suggestion
					resource.TestCheckResourceAttr(data.ResourceName, "additional_capabilities.0.ultra_ssd_enabled", "true"),
```

File: azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go (hunk 1 above)
Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go]: Fixed

File: azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go (hunk 2 above)
Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go]: Fixed

Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go]: Fixed

File: azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go (hunk 3 above)
Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go]: Fixed

File: azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go (hunk 4 above)
Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go]: Fixed

File: azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go (hunk 5 above)
Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go]: Fixed

File: azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go (hunk 6 above)
Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/linux_virtual_machine_resource_other_test.go]: Fixed

File: azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go (hunk 7 above)
Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go]: Fixed

File: azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go (hunk 8 above)
Comment (NilsBusche) [reply azurerm/internal/services/compute/tests/windows_virtual_machine_resource_other_test.go]: Fixed

**Comment by NilsBusche** [comment]
> called out all of the references to be explicit - but if we can fix these test bugs up then this otherwise LGTM 👍

Thanks for this advice. All occurrences should be fixed now.

**Review by tombuildsstuff — APPROVED** [review]
LGTM - thanks for pushing those changes @NilsBusche

**Comment by tombuildsstuff** [comment]
hey @NilsBusche 

Apologies for the delayed re-review here!

I've run the tests for this but it appears the test configurations need to be updated to be deployed into a Zone so that UltraSSD can be used?

> Error: Error creating Linux Virtual Machine "acctestVM-200824104601074216" (Resource Group "acctestRG-200824104601074216"): compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="UltraSSD can be used only with Virtual Machine in an Availability Zone." Target="additionalCapabilities.ultraSSDEnabled"

Thanks!

**Comment by NilsBusche** [comment]
> hey @NilsBusche
> 
> Apologies for the delayed re-review here!
> 
> I've run the tests for this but it appears the test configurations need to be updated to be deployed into a Zone so that UltraSSD can be used?
> 
> > Error: Error creating Linux Virtual Machine "acctestVM-200824104601074216" (Resource Group "acctestRG-200824104601074216"): compute.VirtualMachinesClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="InvalidParameter" Message="UltraSSD can be used only with Virtual Machine in an Availability Zone." Target="additionalCapabilities.ultraSSDEnabled"
> 
> Thanks!

Hi @tombuildsstuff, yes, this is correct. I will add this later. 

But why has the automatic Travis CI been run successfully?

**Comment by tombuildsstuff** [comment]
@NilsBusche Travis only runs the Unit Tests (since these are quick) the Acceptance Tests on the other hand are only run in a separate CI system which since they take longer (and cost money, so we trigger these manually)

**Comment by NilsBusche** [comment]
> @NilsBusche Travis only runs the Unit Tests (since these are quick) the Acceptance Tests on the other hand are only run in a separate CI system which since they take longer (and cost money, so we trigger these manually)

Hi @tombuildsstuff, thanks for this information. I added the `zone` attribute. Hopefully now everything is fine.

**Comment by tombuildsstuff** [comment]
Tests look good - thanks for pushing those changes @NilsBuschel

**Pull request merged**